### PR TITLE
clientv3: use grpc codes to translate raw grpc errors

### PR DESCRIPTION
**Pull request by heyitsanthony**

### CODE REVIEW

File: clientv3/client.go [hunk 1]
```diff
@@ -299,25 +300,34 @@ func isHaltErr(ctx context.Context, err error) bool {
...
-		err = ctx.Err()
-	case strings.Contains(err.Error(), ErrNoAvailableEndpoints.Error()):
+	if _, ok := err.(rpctypes.EtcdError); ok {
+		return err
+	}
+	code := grpc.Code(err)
+	switch code {
+	case codes.DeadlineExceeded:
+		fallthrough
+	case codes.Canceled:
```
Comment (gyuho) [inline clientv3/client.go]: When does this happen?

Comment (heyitsanthony) [inline clientv3/client.go]: should be returned if the rpc call's ctx is canceled

**Comment by gyuho** [comment]
I was debugging unexpected ErrClientConnClosing error from gRPC in txn write fail tests. Would this affect that code path?

**Comment by heyitsanthony** [comment]
Yes, but it's still triggering the failure according to semaphore. I was hoping the tighter error handling in the client would fix the problem.

**Comment by gyuho** [comment]
Yeah this approach looks good. LGTM.

Defer to @xiang90 

Thanks!

File: clientv3/client.go [hunk 2]
```diff
@@ -299,25 +300,34 @@ func isHaltErr(ctx context.Context, err error) bool {
...
+		if ctx.Err() != nil {
+			err = ctx.Err()
+		}
+	case codes.Unavailable:
 		err = ErrNoAvailableEndpoints
-	case strings.Contains(err.Error(), grpc.ErrClientConnClosing.Error()):
+	case codes.FailedPrecondition:
+		fallthrough
+	case codes.Internal:
```
Comment (xiang90) [inline clientv3/client.go]: is this true? all internal errors are errclientconnclosing?

Comment (heyitsanthony) [inline clientv3/client.go]: it seems like the failure path for any internal error should be the same as the client close path, but I can try it without filtering for internal errors

Comment (xiang90) [inline clientv3/client.go]: let's not filter it for now?

Comment (heyitsanthony) [inline clientv3/client.go]: ok fair enough

**Comment by xiang90** [comment]
@heyitsanthony Do we know the root cause of https://github.com/coreos/etcd/issues/6238?

**Comment by xiang90** [comment]
lgtm

File: etcdserver/api/v3rpc/rpctypes/error.go [hunk 3]
```diff
@@ -51,9 +51,11 @@ var (
...
 	ErrGRPCAuthNotEnabled       = grpc.Errorf(codes.FailedPrecondition, "etcdserver: authentication is not enabled")
 
-	ErrGRPCNoLeader   = grpc.Errorf(codes.Unavailable, "etcdserver: no leader")
-	ErrGRPCNotCapable = grpc.Errorf(codes.Unavailable, "etcdserver: not capable")
-	ErrGRPCStopped    = grpc.Errorf(codes.Unavailable, "etcdserver: server stopped")
+	ErrGRPCNoLeader               = grpc.Errorf(codes.Unavailable, "etcdserver: no leader")
+	ErrGRPCNotCapable             = grpc.Errorf(codes.Unavailable, "etcdserver: not capable")
+	ErrGRPCStopped                = grpc.Errorf(codes.Unavailable, "etcdserver: server stopped")
+	ErrGRPCTimeout                = grpc.Errorf(codes.Unavailable, "etcdserver: request timed out")
+	ErrGRPCTimeoutDueToLeaderFail = grpc.Errorf(codes.Unavailable, "etcdserver: request timed out, possibly due to previous leader failure")
```
Comment (xiang90) [inline etcdserver/api/v3rpc/rpctypes/error.go]: where do we use this newly defined error?

Comment (heyitsanthony) [inline etcdserver/api/v3rpc/rpctypes/error.go]: They're initially defined in `etcdserver/errors.go` and bubble up from `etcdserver/server.go` to the `Get` that is timing out on `TestTxnWriteFail` which keeps failing so I can't merge

Comment (xiang90) [inline etcdserver/api/v3rpc/rpctypes/error.go]: can we also update https://github.com/coreos/etcd/blob/master/etcdserver/api/v3rpc/util.go

Comment (heyitsanthony) [inline etcdserver/api/v3rpc/rpctypes/error.go]: sure

**Comment by heyitsanthony** [comment]
Got CI passing. I think this now fixes #6238 PTAL /cc @gyuho

**Comment by gyuho** [comment]
LGTM. Thanks!

**Pull request merged**